### PR TITLE
BitLocker-OpenCL

### PR DESCRIPTION
**Pull request by e-ago**
In this pull request:

- bitlocker2john code reworked (more checks on VMK format)
- bitlocker-opencl:
  - code cleanup 
  - comments in main attack kernel

NB. False positive occurred (see issue opened on my repo: https://github.com/e-ago/bitcracker/issues/1 ); I found an empirically verified and more restrictive check to avoid this type of false positives (-s option, see README https://github.com/e-ago/bitcracker). In the future I will improve this option with a slower but more robust check.
Is it possible to add this optional check in this JtR version? (opening in a different pull request)

### CODE REVIEW

**Comment by kholia** [comment]
Is this specific instance of false positives being found limited to the BitLocker OpenCL format only?

I tested the following hash with the BitLocker CPU format, `$bitlocker$0$16$67337d84e867dab22a1e852e3de6ad2f$1048576$12$a0d228ee8986d20116000000$60$0126c31078113c3c26fe88f161566b0d7b56d4bd8207f635e242bd4e5fdc7790210a7394e5807573306888845f00702d8822dcd0a4398a20839ca2b`

```
$ cat wordlist
password
countosic
alt-csurg
```

CPU format (no false positives in this particular case),

```
$ ../run/john hash -w=wordlist
Warning: detected hash type "BitLocker", but the string is also recognized as "BitLocker-opencl"
Use the "--format=BitLocker-opencl" option to force loading these as that type instead
Using default input encoding: UTF-8
Loaded 1 password hash (BitLocker, BitLocker [SHA-256 AES 64/64])
Cost 1 (iteration count) is 1048576 for all loaded hashes
Will run 4 OpenMP threads
Note: This format may emit false positives, so it will keep trying even after
finding a possible candidate.
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:00 DONE (2017-09-20 09:22) 0g/s 3.333p/s 3.333c/s 3.333C/s password..alt-csurg
Session completed
```

GPU format (has false positives),

```
$ ../run/john --device=0 --format=bitlocker-opencl hash -w=wordlist
Device 0: Intel(R) HD Graphics Skylake ULT GT2
Using default input encoding: UTF-8
Loaded 1 password hash (BitLocker-opencl, BitLocker [SHA256 AES OpenCL])
Cost 1 (iteration count) is 1048576 for all loaded hashes
Note: minimum length forced to 8
Press 'q' or Ctrl-C to abort, almost any other key for status
countosic        (?)
1g 0:00:00:27 DONE (2017-09-20 09:29) 0.03640g/s 0.1092p/s 0.1092c/s 0.1092C/s password..alt-csurg
Use the "--show" option to display all of the cracked passwords reliably
Session completed
```

No false positives were found by the BitLocker CPU format.

I thought that the BitLocker OpenCL format had the same password verification checks as the CPU format.

Also, you should add `FMT_NOT_EXACT` to the `FMT_CASE | FMT_8_BIT` line in opencl_bitlocker_fmt_plug.c file.

**Comment by e-ago** [comment]
@kholia in your CPU format you are doing an additional test about the version (checking bytes 20 and 21 of your output) that I never did in my GPU format. Is this check supported by the standard? If yes, I can add it into my code.

```
if (version == 1 && data_size == 0x2c && v1 <= 0x05 && v2 == 0x20)
```
The version check prevents your CPU format to return a false positive.

The problem of the false positive is the v1 value: for this combination of hash-passwords, v2 and data_size are correct and v1 is between 0x00 and 0x05, then without the version check, my format returns true.
Running several tests with your hashes and my encrypted images (Windows 7, 8.1 and 10), I found that, in case of correct password, v1 is always 0x03. This is the "strict check" I've applied in my version.

File: src/bitlocker2john.c [hunk 1]
```diff
@@ -30,19 +30,44 @@
 #if  (!AC_BUILT || HAVE_UNISTD_H) && !_MSC_VER
 #include <unistd.h> // getopt defined here for unix
 #endif
+#include <getopt.h>
```
[the remark this answers is not in the document]
Comment (e-ago) [reply src/bitlocker2john.c]: Ok, et me know if I have to remove it

**Comment by kholia** [comment]
Thanks for the detailed explanation 👍

I understand the problem better now. I think that the "strict check" should be used by default. Hopefully, it won't lead to any false negatives.

A pull request to add the "strict check" to JtR's BitLocker OpenCL format would be most welcome. Thanks!

**Comment by e-ago** [comment]
The problem is that I can't guarantee that this "strict check" doesn't lead to false negatives (it is better to have some false positive instead! :) ).
For this reason this check is still optional in my repo; I would like to do more tests and research about BitLocker standard.

What about your version check? Is it compliant with the standard? If yes, I could use it too :)

**Comment by kholia** [comment]
I did not discover the extra password verification checks myself. I just ported them from the https://github.com/libyal/libbde/ project to JtR.

According to http://jessekornblum.com/publications/di09.pdf the `version` field is always `1`. So, this check should be safe to use.

Also see https://github.com/libyal/libbde/issues/13.

File: src/bitlocker2john.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (e-ago) [reply src/bitlocker2john.c]: @magnumripper let me know if the new commit with getopt is ok or I have to remove it

**Comment by magnumripper** [comment]
I think getopt is OK.

> The problem is that I can't guarantee that this "strict check" doesn't lead to false negatives (it is better to have some false positive instead! :) ).

👍 We **definitely** don't want false negatives! More research needed!

**Comment by kholia** [comment]
@magnumripper We probably should have used the "Squash and merge" option when merging these commits. The repository history looks a bit "unclean" without squashing in my opinion.

**Comment by magnumripper** [comment]
I did that on purpose, although the commit descriptions were generally poor so it was an edge case.

**Comment by kholia** [comment]
Agreed. Yes, the poor commit descriptions made me complain about the chosen commit strategy.

**Pull request merged**